File: clients/node/src/darwin.ts
import { execFile } from "node:child_process";
import { KeySyncError } from "./index.js";

/** macOS keychain access via the built-in `security` CLI. */
export async function darwinGet(service: string, account: string): Promise<string> {
  const { stdout, stderr } = await execFilePromise("security", [
    "find-generic-password",
    "-s", service,
    "-a", account,
    "-w",
  ]);
  if (stderr && !stdout) {
    throw new KeySyncError("notFound", `secret not found: ${service}/${account}`);
  }
  return stdout.trim();
}

/** List all keysync secrets on macOS by parsing `security dump-keychain`. */
export async function darwinList(): Promise<Array<{ service: string; account: string }>> {
  const { stdout } = await execFilePromise("security", ["dump-keychain"]);
  const records = stdout.split("\nkeychain:");
  const results: Array<{ service: string; account: string }> = [];

  for (const rec of records) {
    if (!rec.includes(`class: "genp"`)) continue;
    const svc = findAttrValue(rec, "svce");
    if (!svc?.startsWith("keysync/")) continue;
    const acct = findAttrValue(rec, "acct");
    if (acct) {
      results.push({ service: svc, account: acct });
    }
  }
  return results;
}

function findAttrValue(record: string, attrName: string): string | undefined {
  const idx = record.indexOf(`"${attrName}"`);
  if (idx < 0) return undefined;
  const after = record.slice(idx + attrName.length + 2);
  const eqIdx = after.indexOf("=");
  if (eqIdx < 0) return undefined;
  let val = after.slice(eqIdx + 1).trim();
  if (val === "<NULL>") return undefined;
  if (val.startsWith('"')) {
    const end = val.indexOf('"', 1);
    return end >= 0 ? val.slice(1, end) : val.slice(1);
  }
  return val.replace(/^"|"$/g, "");
}

/** Wrapper around child_process.execFile that returns a promise. */
function execFilePromise(
  cmd: string,
  args: string[]
): Promise<{ stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, (error, stdout, stderr) => {
      if (error) {
        // security returns exit code 44 for "item not found"
        if ((error as NodeJS.ErrnoException).code === 44) {
          resolve({ stdout: "", stderr: "not found" });
        } else {
          reject(error);
        }
      } else {
        resolve({ stdout, stderr });
      }
    });
  });
}

File: clients/go/darwin.go
//go:build darwin

package keysync

import (
	"fmt"
	"os/exec"
	"strings"
)

func init() {
	isNotFound = darwinIsNotFound
	platformGet = darwinGet
	platformList = darwinList
}

// darwinGet retrieves a secret from the macOS Keychain via the security CLI.
func darwinGet(service, account string) (string, error) {
	cmd := exec.Command("security", "find-generic-password",
		"-s", service,
		"-a", account,
		"-w",
	)
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("security find-generic-password: %w", err)
	}
	return strings.TrimSpace(string(out)), nil
}

// darwinList lists all keysync secrets by scanning the keychain.
func darwinList(scope, project, environment string) ([]SecretEntry, error) {
	cmd := exec.Command("security", "dump-keychain")
	out, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("security dump-keychain: %w", err)
	}

	records := strings.Split(string(out), "\nkeychain:")
	var entries []SecretEntry
	for _, rec := range records {
		rec = strings.TrimSpace(rec)
		if rec == "" || !strings.Contains(rec, `class: "genp"`) {
			continue
		}
		svc := findAttrValue(rec, "svce")
		if !strings.HasPrefix(svc, "keysync/") {
			continue
		}
		entryScope, entryProject, entryEnvironment := parseServiceName(svc)
		acct := findAttrValue(rec, "acct")
		if acct == "" {
			continue
		}
		if (scope == "" || entryScope == scope) &&
			(project == "" || entryProject == project) &&
			(environment == "" || entryEnvironment == environment) {
			entries = append(entries, SecretEntry{
				Scope:       entryScope,
				Project:     entryProject,
				Environment: entryEnvironment,
				Key:         acct,
			})
		}
	}
	return entries, nil
}

func darwinIsNotFound(err error) bool {
	if exitErr, ok := err.(*exec.ExitError); ok {
		return exitErr.ExitCode() == 44 // security returns 44 for item not found
	}
	return false
}

// findAttrValue extracts an attribute value from a dump-keychain record.
func findAttrValue(record, attrName string) string {
	idx := strings.Index(record, fmt.Sprintf(`"%s"`, attrName))
	if idx < 0 {
		return ""
	}
	after := record[idx+len(attrName)+2:]
	eqIdx := strings.IndexByte(after, '=')
	if eqIdx < 0 {
		return ""
	}
	val := strings.TrimSpace(after[eqIdx+1:])
	if val == "<NULL>" {
		return ""
	}
	if strings.HasPrefix(val, `"`) {
		end := strings.IndexByte(val[1:], '"')
		if end >= 0 {
			return val[1 : end+1]
		}
	}
	return strings.Trim(val, `"`)
}
